include/wil/safearrays.h

```diff
@@ -0,0 +1,897 @@
+    //!         DoSomethingWithBSTR(bstr);
+    //!     }
+    //!     return S_OK;
+    //! }
+    //! ~~~~
+    //! @tparam storage_t   Storage class that manages the SAFEARRAY pointer
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    //! @tparam element_t   The underlying datatype that is contained in the safearray OR
+    //!                     it can be void to make the safearray generic (and not as typesafe)
+    template <typename storage_t, typename err_policy = err_exception_policy, typename element_t = void>
+    class safearray_t : public storage_t
+    {
+    private:
+        // SFINAE Helpers to improve readability
+        template<typename T>
+        using is_pointer_type = wistd::disjunction<wistd::is_same<T, BSTR>,
+                                                    wistd::is_same<T, LPUNKNOWN>,
+                                                    wistd::is_same<T, LPDISPATCH>>;
+        template<typename T>
+        using is_type_not_set = wistd::is_same<T, void>;
+        template<typename T>
+        using is_type_set = wistd::negation<is_type_not_set<T>>;
+        template <typename T>
+        using is_value_type = wistd::conjunction<wistd::negation<is_pointer_type<T>>, is_type_set<T>>;
+
+    public:
+        template<typename T>
+        using safearraydata_t = safearraydata_t<T, err_policy>;
+
+        // Represents this type
+        using unique_type = unique_any_t<safearray_t<storage_t, err_policy, element_t>>;
+
+        typedef typename err_policy::result result;
+        typedef typename storage_t::pointer pointer;
+        typedef element_t elemtype;
+
+    public:
+        // forward all base class constructors...
+        template <typename... args_t>
+        explicit safearray_t(args_t&&... args) WI_NOEXCEPT : storage_t(wistd::forward<args_t>(args)...) {}
+
+        // Exception-based construction
+        template<typename T = element_t, typename wistd::enable_if<is_type_not_set<T>::value, int>::type = 0>
+        safearray_t(VARTYPE vt, UINT cElements, LONG lowerBound = 0)
+        {
+            static_assert(wistd::is_same<void, result>::value, "this constructor requires exceptions; use the create method");
+            create(vt, cElements, lowerBound);
+        }
+        template<typename T = element_t, typename wistd::enable_if<is_type_set<T>::value, int>::type = 0>
+        safearray_t(UINT cElements, LONG lowerBound = 0)
+        {
+            static_assert(wistd::is_same<void, result>::value, "this constructor requires exceptions; use the create method");
+            create<T>(cElements, lowerBound);
+        }
+
+        // Low-level, arbitrary number of dimensions
+        // Uses Explicit Type
+        template<typename T = element_t, typename wistd::enable_if<is_type_not_set<T>::value, int>::type = 0>
+        result create(VARTYPE vt, UINT cDims, SAFEARRAYBOUND* sab)

```

Comment:
I'd say follow STL naming convention and call these `reset`

Reply:
I'm not sure about this... In STL that means you giving it a new pointer that you want to reset the smart pointer to.  These functions take parameters to create a new instance of the object.

Let's see if anyone else has an opinion on this because I'm on the fence.

Reply:
I suppose `assign` would perhaps be a better option